lightning-net-tokio/src/lib.rs

```diff
@@ -1,121 +1,179 @@
+									lock.read_paused = true;
+								}
 
-			future::Either::B(future::result(Ok(())))
-		}).then(move |_| {
-			if us_close_ref.lock().unwrap().need_disconnect {
-				peer_manager_ref.disconnect_event(&SocketDescriptor::new(us_close_ref, peer_manager_ref.clone()));
-				println!("Peer disconnected!");
-			} else {
-				println!("We disconnected peer!");
+								match us.lock().unwrap().event_notify.try_send(()) {
+									Ok(_) => {},
+									Err(mpsc::error::TrySendError::Full(_)) => {
+										// Ignore full errors as we just need them to poll after this point, so if the user
+										// hasn't received the last send yet, it doesn't matter.
+									},
+									_ => panic!()
+								}
+							},
+							Err(e) => shutdown_socket!(e),
+						}
+					},
+					Err(e) => {
+						println!("Connection closed: {}", e);
+						break;
+					},
+				},
+			}
+		}
+		let writer_option = us.lock().unwrap().writer.take();
+		if let Some(mut writer) = writer_option {
+			writer.shutdown().await.expect("We should be able to shutdown() a socket, even if it is already disconnected");
+		}
+		if us.lock().unwrap().need_disconnect_event {
+			peer_manager_ref.disconnect_event(&SocketDescriptor::new(Arc::clone(&us)));
+			match us.lock().unwrap().event_notify.try_send(()) {
+				Ok(_) => {},
+				Err(mpsc::error::TrySendError::Full(_)) => {
+					// Ignore full errors as we just need them to poll after this point, so if the user
+					// hasn't received the last send yet, it doesn't matter.
+				},
+				_ => panic!()
 			}
-			Ok(())
-		}));
+		}
 	}
 
-	fn new(event_notify: mpsc::Sender<()>, stream: TcpStream) -> (futures::stream::SplitStream<tokio_codec::Framed<TcpStream, tokio_codec::BytesCodec>>, Arc<Mutex<Self>>) {
-		let (writer, reader) = tokio_codec::Framed::new(stream, tokio_codec::BytesCodec::new()).split();
-		let (send_sink, send_stream) = mpsc::channel(3);
-		tokio::spawn(writer.send_all(send_stream.map_err(|_| -> std::io::Error {
-			unreachable!();
-		})).then(|_| {
-			future::result(Ok(()))
-		}));
-		let us = Arc::new(Mutex::new(Self { writer: Some(send_sink), event_notify, pending_read: Vec::new(), read_blocker: None, read_paused: false, need_disconnect: true, id: ID_COUNTER.fetch_add(1, Ordering::AcqRel) }));
+	fn new(event_notify: mpsc::Sender<()>, stream: TcpStream) -> (io::ReadHalf<TcpStream>, mpsc::Receiver<()>, Arc<Mutex<Self>>) {
+		// We only ever need a channel of depth 1 here: if we returned a non-full write to the
+		// PeerManager, we will eventually get notified that there is room in the socket to write
+		// new bytes, which will generate an event. That event will be popped off the queue before
+		// we call write_event, ensuring that we have room to push a new () if, during the
+		// write_event() call, send_data() returns a non-full write.
+		let (write_avail, receiver) = mpsc::channel(1);
+		let (reader, writer) = io::split(stream);
 
-		(reader, us)
+		(reader, receiver,
+		Arc::new(Mutex::new(Self {
```

Comment:
I'm not quite sure I follow why an `Arc` and a `Mutex` are needed. Could you describe a scenario for me where a Connection may be accessed concurrently?